Python UDF with arrow type incorrectly returns the cache result of the first row

### What happens?

When defining an Python UDF with `type="arrow"`, e.g. 
```
import random as r

def random_arrow(x):
        return [r.randint(0, 10) for item in x.to_pylist()]
...
duckdb.create_function(
        "random_arrow",
        random_arrow,
        [VARCHAR],
        INTEGER,
        side_effects=True,
        type="arrow",
    )
```
executing a query which call the UDF with the same argument results in incorrect results (which contains the first "row" duplicated up to results set size), i.e.
```sql
SELECT random_arrow('') FROM range(10)
```
returns `[(1,), (1), (1,), ...]` instead of random numbers.
This incorrect and doesn't seems to honor the `side_effects=True` flag.

Note: Using `SELECT random_arrow(range) FROM range(10)` returns correct results.

Attached is a minimally reproducing test.




### To Reproduce

```
import duckdb
import random as r

from duckdb.typing import INTEGER, VARCHAR

def random_arrow(x):
    return [r.randint(0, 10) for item in x.to_pylist()]

def random_non_arrow(x):
    return r.randint(0, 10)

con = duckdb.connect()

con.create_function(
    "random_non_arrow",
    random_non_arrow,
    [VARCHAR],
    INTEGER,
    side_effects=True,
)

con.create_function(
    "random_arrow",
    random_arrow,
    [VARCHAR],
    INTEGER,
    side_effects=True,
    type="arrow",
)

res = con.sql("select random_non_arrow('') from range(10)").fetchall()
assert len(set(res)) > 1

res = con.sql("select random_arrow(range) from range(10)").fetchall()
assert len(set(res)) > 1

res = con.sql("select random_arrow('') from range(10)").fetchall()
assert len(set(res)) > 1 # should pass
```
fails 
```
>       assert len(set(res)) > 1
E       assert 1 > 1
E        +  where 1 = len({(10,)})
E        +    where {(10,)} = set([(10,), (10,), (10,), (10,), (10,), (10,), ...])
```


### OS:

arm64

### DuckDB Version:

0.9.2

### DuckDB Client:

Python

### Full Name:

Chi Lang Ngo

### Affiliation:

NA

### Have you tried this on the latest `main` branch?

I have tested with a release build (and could not test with a main build)

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

--- a/tools/pythonpkg/scripts/cache_data.json
+++ b/tools/pythonpkg/scripts/cache_data.json
@@ -214,6 +214,7 @@
         "full_path": "numpy",
         "name": "numpy",
         "children": [
+			"numpy.core",
             "numpy.ndarray",
             "numpy.datetime64",
             "numpy.generic",
@@ -237,6 +238,20 @@
             "numpy.cdouble",
             "numpy.clongdouble"
         ]
+    },
+    "numpy.core": {
+        "type": "attribute",
+        "full_path": "numpy.core",
+        "name": "core",
+        "children": [
+            "numpy.core.multiarray"
+        ]
+    },
+    "numpy.core.multiarray": {
+        "type": "attribute",
+        "full_path": "numpy.core.multiarray",
+        "name": "multiarray",
+        "children": []
     },
     "numpy.ndarray": {
         "type": "attribute",

--- a/tools/pythonpkg/scripts/imports.py
+++ b/tools/pythonpkg/scripts/imports.py
@@ -42,6 +42,7 @@ ipywidgets.FloatProgress
 
 import numpy
 
+numpy.core.multiarray
 numpy.ndarray
 numpy.datetime64
 numpy.generic

--- a/tools/pythonpkg/src/include/duckdb_python/import_cache/modules/numpy_module.hpp
+++ b/tools/pythonpkg/src/include/duckdb_python/import_cache/modules/numpy_module.hpp
@@ -13,6 +13,18 @@
 
 namespace duckdb {
 
+struct NumpyCoreCacheItem : public PythonImportCacheItem {
+
+public:
+	NumpyCoreCacheItem(optional_ptr<PythonImportCacheItem> parent)
+	    : PythonImportCacheItem("core", parent), multiarray("multiarray", this) {
+	}
+	~NumpyCoreCacheItem() override {
+	}
+
+	PythonImportCacheItem multiarray;
+};
+
 struct NumpyCacheItem : public PythonImportCacheItem {
 
 public:
@@ -20,7 +32,7 @@ public:
 
 public:
 	NumpyCacheItem()
-	    : PythonImportCacheItem("numpy"), ndarray("ndarray", this), datetime64("datetime64", this),
+	    : PythonImportCacheItem("numpy"), core(this), ndarray("ndarray", this), datetime64("datetime64", this),
 	      generic("generic", this), int64("int64", this), bool_("bool_", this), byte("byte", this),
 	      ubyte("ubyte", this), short_("short", this), ushort_("ushort", this), intc("intc", this),
 	      uintc("uintc", this), int_("int_", this), uint("uint", this), longlong("longlong", this),
@@ -31,6 +43,7 @@ public:
 	~NumpyCacheItem() override {
 	}
 
+	NumpyCoreCacheItem core;
 	PythonImportCacheItem ndarray;
 	PythonImportCacheItem datetime64;
 	PythonImportCacheItem generic;

--- a/tools/pythonpkg/src/python_udf.cpp
+++ b/tools/pythonpkg/src/python_udf.cpp
@@ -147,7 +147,7 @@ static scalar_function_t CreateVectorizedFunction(PyObject *function, PythonExce
 		// Convert the pyarrow result back to a DuckDB datachunk
 		ConvertPyArrowToDataChunk(python_object, result, state.GetContext(), count);
 
-		if (input.AllConstant()) {
+		if (input.size() == 1) {
 			result.SetVectorType(VectorType::CONSTANT_VECTOR);
 		}
 	};
@@ -322,6 +322,11 @@ public:
 
 	ScalarFunction GetFunction(const py::function &udf, PythonExceptionHandling exception_handling, bool side_effects,
 	                           const ClientProperties &client_properties) {
+
+		auto &import_cache = *DuckDBPyConnection::ImportCache();
+		// Import this module, because importing this from a non-main thread causes a segfault
+		(void)import_cache.numpy.core.multiarray();
+
 		scalar_function_t func;
 		if (vectorized) {
 			func = CreateVectorizedFunction(udf.ptr(), exception_handling);

--- a/tools/pythonpkg/tests/fast/udf/test_scalar_arrow.py
+++ b/tools/pythonpkg/tests/fast/udf/test_scalar_arrow.py
@@ -128,13 +128,13 @@ class TestPyArrowUDF(object):
         import random as r
 
         def random_arrow(x):
-            if 'data' not in random_arrow.__dict__:
-                random_arrow.__dict__["data"] = 0
+            if not hasattr(random_arrow, 'data'):
+                random_arrow.data = 0
 
             input = x.to_pylist()
-            val = random_arrow.__dict__["data"]
+            val = random_arrow.data
             output = [val + i for i in range(len(input))]
-            random_arrow.__dict__["data"] += len(input)
+            random_arrow.data += len(input)
             return output
 
         duckdb_cursor.create_function(

--- a/tools/pythonpkg/tests/fast/udf/test_scalar_arrow.py
+++ b/tools/pythonpkg/tests/fast/udf/test_scalar_arrow.py
@@ -124,6 +124,30 @@ class TestPyArrowUDF(object):
         with pytest.raises(duckdb.InvalidInputException, match='Returned pyarrow table should have 1 tuples, found 5'):
             res = con.sql("""select too_many_tuples(5)""").fetchall()
 
+    def test_arrow_side_effects(self, duckdb_cursor):
+        import random as r
+
+        def random_arrow(x):
+            if 'data' not in random_arrow.__dict__:
+                random_arrow.__dict__["data"] = 0
+
+            input = x.to_pylist()
+            val = random_arrow.__dict__["data"]
+            output = [val + i for i in range(len(input))]
+            random_arrow.__dict__["data"] += len(input)
+            return output
+
+        duckdb_cursor.create_function(
+            "random_arrow",
+            random_arrow,
+            [VARCHAR],
+            INTEGER,
+            side_effects=True,
+            type="arrow",
+        )
+        res = duckdb_cursor.query("SELECT random_arrow('') FROM range(10)").fetchall()
+        assert res == [(0,), (1,), (2,), (3,), (4,), (5,), (6,), (7,), (8,), (9,)]
+
     def test_return_struct(self):
         def return_struct(col):
             con = duckdb.connect()

